src/stages/mining_create_block.rs

```diff
@@ -0,0 +1,172 @@
+use num_traits::ToPrimitive;
+use std::{
+    rc::Rc,
+    sync::{mpsc, Arc, Mutex},
+};
+use tokio::io::copy;
+use tracing::debug;
+
+pub const CREATE_BLOCK: StageId = StageId("CreateBlock");
+// DAOForkExtraRange is the number of consecutive blocks from the DAO fork point
+// to override the extra-data in to prevent no-fork attacks.
+pub const DAOFORKEXTRARANG: i32 = 10;
+
+#[derive(Debug)]
+pub struct MiningBlock {
+    pub header: BlockHeader,
+    pub ommers: Vec<BlockHeader>,
+    pub transactions: Vec<MessageWithSignature>,
+}
+
+impl MiningStatus {
+    pub fn new() -> Self {
+        Self {
+            pending_result_ch: mpsc::channel().0,
+            mining_result_ch: mpsc::channel().0,
+            mining_result_pos_ch: mpsc::channel().0,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct CreateBlock {
+    pub mining_status: Arc<Mutex<MiningStatus>>,
+    pub mining_block: Arc<Mutex<MiningBlock>>,
+    pub mining_config: Arc<Mutex<MiningConfig>>,
+    pub chain_spec: ChainSpec,
+}
+
+#[async_trait]
+impl<'db, E> Stage<'db, E> for CreateBlock
+where
+    E: EnvironmentKind,
+{
+    fn id(&self) -> StageId {
+        CREATE_BLOCK
+    }
+
+    async fn execute<'tx>(
+        &mut self,
+        tx: &'tx mut MdbxTransaction<'db, RW, E>,
+        input: StageInput,
+    ) -> Result<ExecOutput, StageError>
+    where
+        'db: 'tx,
+    {
+        let mining_block = MiningBlock {
+            header: BlockHeader {
+                parent_hash: H256::zero(),
+                ommers_hash: H256::zero(),
+                beneficiary: Address::zero(),
+                state_root: H256::zero(),
+                transactions_root: H256::zero(),
+                receipts_root: H256::zero(),
+                logs_bloom: Bloom::zero(),
+                difficulty: U256::ZERO,
+                number: BlockNumber(0),
+                gas_limit: 0,
+                gas_used: 0,
+                timestamp: 0,
+                extra_data: Bytes::new(),
+                mix_hash: H256::zero(),
+                nonce: H64::zero(),
+                base_fee_per_gas: None,
+            },
+            ommers: vec![],
+            transactions: vec![],
+        };
+        let mining_block_mutex = Arc::new(Mutex::new(mining_block));
+        self.mining_block = Arc::clone(&mining_block_mutex);
+        let parent_number = input.stage_progress.unwrap();
+        let parent_header = get_header(tx, parent_number)?;
+
+        // TODO, complete the remaining tx related part after txpool ready.
+        let mut proposal = create_block_header(&parent_header, Arc::clone(&self.mining_config))?;
+        debug!("Empty block created: {:?}", proposal);
+
+        let buffer = Buffer::new(tx, None);
+        if self.chain_spec.consensus.is_parlia() {
+            assert!(self
+                .mining_config
+                .lock()
+                .unwrap()
+                .consensus
+                .prepare(&buffer, &mut proposal)
```

Comment:
before you call prepare function, you must call snapshot first, it require snaps.